### PR TITLE
add SUSE/openSUSE

### PR DESCRIPTION
**Pull request by johanneskastl**
Similar to the PR for the targetcli role.

Currently only tested on Leap 15.1, but I will try to test it with at least SLES15 SP1 and SLES12 SP4 this week.

### CODE REVIEW

**Comment by OndrejHome** [comment]
Hi @johanneskastl, Thank you for also this one. I will also check this one out over week. I plan to test it on same systems as I have done the targetcli checks. Once confirmed on which systems this works well the `meta/main.yml` would also need to be adjusted to reflect that.

File: tasks/main.yml [hunk 1]
```diff
@@ -1,12 +1,19 @@
 ---
-- name: Install packages
+
+- include_vars: vars_RedHat_CentOS.yml
+  when: ansible_distribution == 'RedHat' or ansible_distribution == 'CentOS'
+
+- include_vars: vars_SUSE_openSUSE.yml
+  when: 'ansible_os_family == "Suse" or ansible_os_family == "openSUSE Leap"'
```
Comment (OndrejHome) [inline tasks/main.yml:7]: Same as in the PR for targetcli I would request the move of `vars_*` files from `tasks/` directory to `vars/` directory - https://github.com/OndrejHome/ansible.targetcli/pull/4#issuecomment-603202848

File: tasks/main.yml [hunk 2]
```diff
@@ -18,7 +25,11 @@
     dest: '/etc/iscsi/initiatorname.iscsi'
     regexp: '^InitiatorName='
     line: "InitiatorName={{ custom_iscsi_initiator_name is defined|ternary(custom_iscsi_initiator_name,iscsi_initiator_name) }}"
-  notify: restart iscsid
+    create: 'yes'
+    owner: 'root'
+    group: 'root'
+    mode: '0600'
```
Comment (OndrejHome) [inline tasks/main.yml:31]: **[?]** Is there reason why the this file should be created? On which OS/package version this file was not present so  `create: yes` option was needed?

**Review by OndrejHome — CHANGES_REQUESTED** [review]
To summarize the request for this PR to proceed:
- (medium) move of `vars*` files  to `vars/` directory
- (medium) reasoning for `create: yes` option to `lineinfile` module
- (medium) addition of supported OS versions to `meta/main.yml` file - these should be versions that this module was tested with (I have ability to check compatibility on SLES 12/15 and openSUSE 15 at this moment)

File: tasks/main.yml (hunk 2 above)
Comment (johanneskastl) [reply tasks/main.yml:31]: I had a case where I removed /etc/iscsi, reinstalled all packages and that file was somehow missing.
Might not be needed too often, but makes this less errorprone.

File: tasks/main.yml (hunk 1 above)
Comment (johanneskastl) [reply tasks/main.yml:7]: Sure, will do.

**Comment by johanneskastl** [comment]
Ondrej, I believe the library fails if there is more than one line inside the /etc/iscsi/initatorname.iscsi:

```
The full traceback is:
Traceback (most recent call last):
  File "/home/kastl/.ansible/tmp/ansible-tmp-1585340347.094774-242513872061225/AnsiballZ_iscsiadm_facts.py", line 102, in <module>
    _ansiballz_main()
  File "/home/kastl/.ansible/tmp/ansible-tmp-1585340347.094774-242513872061225/AnsiballZ_iscsiadm_facts.py", line 94, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/home/kastl/.ansible/tmp/ansible-tmp-1585340347.094774-242513872061225/AnsiballZ_iscsiadm_facts.py", line 40, in invoke_module
    runpy.run_module(mod_name='ansible.modules.iscsiadm_facts', init_globals=None, run_name='__main__', alter_sys=True)
  File "/usr/lib64/python2.7/runpy.py", line 188, in run_module
    fname, loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 82, in _run_module_code
    mod_name, mod_fname, mod_loader, pkg_name)
  File "/usr/lib64/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/tmp/ansible_iscsiadm_facts_payload_4vshX0/ansible_iscsiadm_facts_payload.zip/ansible/modules/iscsiadm_facts.py", line 44, in <module>
  File "/tmp/ansible_iscsiadm_facts_payload_4vshX0/ansible_iscsiadm_facts_payload.zip/ansible/modules/iscsiadm_facts.py", line 35, in main
IndexError: list index out of range
fatal: [kastl-SLES12SP4-suseha-client01]: FAILED! => {
    "changed": false,
    "module_stderr": "OpenSSH_8.1p1, OpenSSL 1.1.1d  10 Sep 2019\r\ndebug1: Reading configuration data /home/kastl/.ssh/config\r\ndebug1: /home/kastl/.ssh/config line 138: Applying options for 172.30.*\r\ndebug1: /home/kastl/.ssh/config line 148: Applying options for *\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 20: Applying options for *\r\ndebug2: resolve_canonicalize: hostname 172.30.0.77 is address\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 29693\r\ndebug3: mux_client_request_session: session request sent\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\nShared connection to 172.30.0.77 closed.\r\n",
    "module_stdout": "Traceback (most recent call last):\r\n  File \"/home/kastl/.ansible/tmp/ansible-tmp-1585340347.094774-242513872061225/AnsiballZ_iscsiadm_facts.py\", line 102, in <module>\r\n    _ansiballz_main()\r\n  File \"/home/kastl/.ansible/tmp/ansible-tmp-1585340347.094774-242513872061225/AnsiballZ_iscsiadm_facts.py\", line 94, in _ansiballz_main\r\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\r\n  File \"/home/kastl/.ansible/tmp/ansible-tmp-1585340347.094774-242513872061225/AnsiballZ_iscsiadm_facts.py\", line 40, in invoke_module\r\n    runpy.run_module(mod_name='ansible.modules.iscsiadm_facts', init_globals=None, run_name='__main__', alter_sys=True)\r\n  File \"/usr/lib64/python2.7/runpy.py\", line 188, in run_module\r\n    fname, loader, pkg_name)\r\n  File \"/usr/lib64/python2.7/runpy.py\", line 82, in _run_module_code\r\n    mod_name, mod_fname, mod_loader, pkg_name)\r\n  File \"/usr/lib64/python2.7/runpy.py\", line 72, in _run_code\r\n    exec code in run_globals\r\n  File \"/tmp/ansible_iscsiadm_facts_payload_4vshX0/ansible_iscsiadm_facts_payload.zip/ansible/modules/iscsiadm_facts.py\", line 44, in <module>\r\n  File \"/tmp/ansible_iscsiadm_facts_payload_4vshX0/ansible_iscsiadm_facts_payload.zip/ansible/modules/iscsiadm_facts.py\", line 35, in main\r\nIndexError: list index out of range\r\n",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
 
```

Content of /etc/iscsi/initiatorname.iscsi on SLES12 SP4 directly after installation:
```
##
## /etc/iscsi/iscsi.initiatorname
##
## Default iSCSI Initiatorname.
##
## DO NOT EDIT OR REMOVE THIS FILE! 
## If you remove this file, the iSCSI daemon will not start.
## If you change the InitiatorName, existing access control lists
## may reject this initiator.  The InitiatorName must be unique
## for each iSCSI initiator.  Do NOT duplicate iSCSI InitiatorNames.
InitiatorName=iqn.1996-04.de.suse:01:ae898e32a453
```

Manually edting this to remove the comments:
```
InitiatorName=iqn.1996-04.de.suse:01:ae898e32a453
```

Then the role in its current state is working until the `Login into iSCSI server` task (that does nothing for some reason).

**Comment by johanneskastl** [comment]
> Ondrej, I believe the library fails if there is more than one line inside the /etc/iscsi/initatorname.iscsi:

Fixed in commit ec3f843 that ignores lines starting with a hash (i.e. comments). Please feel free to suggest a better solution, I am not a good python programmer.

**Comment by johanneskastl** [comment]
> Then the role in its current state is working until the `Login into iSCSI server` task (that does nothing for some reason).

`"msg": "/var/lib/iscsi/send_targets/10.250.5.83,3260 was skipped as it does not seem to be a valid directory or it cannot be accessed\n"`

Seems like there is a difference between CentOS and SLES/openSUSE. I need to dig deeper where this is stored and how to find out that information.

**Comment by johanneskastl** [comment]
> Seems like there is a difference between CentOS and SLES/openSUSE. I need to dig deeper where this is stored and how to find out that information.

Found it, see https://github.com/OndrejHome/ansible.iscsiadm/pull/1/commits/7ad415c6b1105a9eaf837609c5f753d73835b601

**Comment by johanneskastl** [comment]
@OndrejHome Is the last task "Login into iSCSI server" idempotent on CentOS?

Because this fails on second run of my playbook:
```
failed: [kastl-SLES15SP1-suseha-client01] (item={'path': '/etc/iscsi/send_targets/10.250.5.78,3260/iqn.2020-03.de.b1-systems.training:target,10.250.5.78,3260,1,default', 'mode': '0777', 'isdir': False, 'ischr': False, 'isblk': False, 'isreg': False, 'isfifo': False, 'islnk': True, 'issock': False, 'uid': 0, 'gid': 0, 'size': 77, 'inode': 84410629, 'dev': 2051, 'nlink': 1, 'atime': 1585341997.2911773, 'mtime': 1585341995.4751596, 'ctime': 1585341995.4751596, 'gr_name': 'root', 'pw_name': 'root', 'wusr': True, 'rusr': True, 'xusr': True, 'wgrp': True, 'rgrp': True, 'xgrp': True, 'woth': True, 'roth': True, 'xoth': True, 'isuid': False, 'isgid': False}) => {"ansible_loop_var": "item", "changed": true, "cmd": ["iscsiadm", "--mode", "node", "--targetname", "iqn.2020-03.de.b1-systems.training:target", "--login", "--portal", "10.250.5.78:3260"], "delta": "0:00:00.006692", "end": "2020-03-27 20:46:38.879662", "item": {"atime": 1585341997.2911773, "ctime": 1585341995.4751596, "dev": 2051, "gid": 0, "gr_name": "root", "inode": 84410629, "isblk": false, "ischr": false, "isdir": false, "isfifo": false, "isgid": false, "islnk": true, "isreg": false, "issock": false, "isuid": false, "mode": "0777", "mtime": 1585341995.4751596, "nlink": 1, "path": "/etc/iscsi/send_targets/10.250.5.78,3260/iqn.2020-03.de.b1-systems.training:target,10.250.5.78,3260,1,default", "pw_name": "root", "rgrp": true, "roth": true, "rusr": true, "size": 77, "uid": 0, "wgrp": true, "woth": true, "wusr": true, "xgrp": true, "xoth": true, "xusr": true}, "msg": "non-zero return code", "rc": 15, "start": "2020-03-27 20:46:38.872970", "stderr": "iscsiadm: default: 1 session requested, but 1 already present.\niscsiadm: Could not log into all portals", "stderr_lines": ["iscsiadm: default: 1 session requested, but 1 already present.", "iscsiadm: Could not log into all portals"], "stdout": "", "stdout_lines": []}
```

Other than that the role seems to work fine.

**Comment by johanneskastl** [comment]
> Found it, see [7ad415c](https://github.com/OndrejHome/ansible.iscsiadm/commit/7ad415c6b1105a9eaf837609c5f753d73835b601)

Refactored that a little, to avoid "Discover WWNs from iSCSI server" being shown as "Changed" on SLES/openSUSE

File: library/iscsiadm_facts.py [hunk 3]
```diff
@@ -30,7 +30,10 @@ def main():
 
         try:
             file = open('/etc/iscsi/initiatorname.iscsi', 'r')
-            text = file.readlines()[0].rstrip()
+            for line in file:
+                if line.startswith('#'):
+                    continue
+                text = line.rstrip()
```
Comment (OndrejHome) [inline library/iscsiadm_facts.py:36]: good job! 
I was not considering that this file could contain comments as it is usually a single line on CentOS/RHEL systems.

File: tasks/main.yml [hunk 4]
```diff
@@ -1,12 +1,19 @@
...
-- name: Install packages
+
+- include_vars: vars_RedHat_CentOS.yml
+  when: ansible_distribution == 'RedHat' or ansible_distribution == 'CentOS'
+
+- include_vars: vars_SUSE_openSUSE.yml
+  when: 'ansible_os_family == "Suse" or ansible_os_family == "openSUSE Leap"'
+
+- name: "Install package(s) {{ iscsiadmin_package }}"
   yum:
```
Comment (OndrejHome) [inline tasks/main.yml]: Is this really working on SUSE with 'yum' module?

File: tasks/main.yml (hunk 2 above)
Comment (OndrejHome) [reply tasks/main.yml:31]: As I don't see any nice way how to force reinstallation of package that would bring this back reliably lets keep this.

**Review by OndrejHome — CHANGES_REQUESTED** [review]
Hi @johanneskastl ,

In overall this looks good.  I have not tested this yet on my testing systems, but I'm planning to get into that over next week.

'Login to iSCSI server' is not idempotent in all situation and I'm aware of that situation as could be seen from comment here - https://github.com/OndrejHome/ansible.iscsiadm/blob/master/tasks/connect_to_target.yml#L15-L16. Unless you feel this is easy-to-fix I would for now avoid changing it in this PR.

(medium) One thing that is missing here is the change to `meta/main.yml` to indicate that openSUSE/SLES are now supported. 

Additional comments in code.

File: tasks/main.yml (hunk 4 above)
Comment (johanneskastl) [reply tasks/main.yml]: It is apparently. Nevertheless I have changed that to `package` now.

**Comment by johanneskastl** [comment]
> Hi @johanneskastl ,
> 
> In overall this looks good. I have not tested this yet on my testing systems, but I'm planning to get into that over next week.

The role is now in a state that is working for me on
- SLES12 SP4
- SLES15 SP1
- openSUSE Leap 15.1

Feel free to report back any issues.
 
> 'Login to iSCSI server' is not idempotent in all situation and I'm aware of that situation as could be seen from comment here - https://github.com/OndrejHome/ansible.iscsiadm/blob/master/tasks/connect_to_target.yml#L15-L16. Unless you feel this is easy-to-fix I would for now avoid changing it in this PR.

Not off the top of my head, unfortunately. I just wanted to make sure that this error has not been introduced by my modifications.

> (medium) One thing that is missing here is the change to `meta/main.yml` to indicate that openSUSE/SLES are now supported.

Has been added, as well as the note in README regarding the SLES/openSUSE 15.x python interpreter.

**Comment by OndrejHome** [comment]
Hi @johanneskastl ,

Thank you for the changes. Checking the code it looks good to me but I still want to test this myself before merging it. I expect to get to this over this week. Thank you for your patience.

**Comment by OndrejHome** [comment]
Hi @johanneskastl , I have tested the changes on following systems without any major issues:
SLES12, SLES15, OpenSUSE 15, CentOS 7.
There was one strange behaviour that I have hit on SLES12 and SLES15 during discovery. It seems that during first attempt of discovery the symlinks in `/etc/iscsi/send_targets/192.168.../` don't get created. I was able to get them created on second attempt by manually running discovery. However then after removing all related files to discovery this started to work normally/as expected.
I didn't expected this to be a problem for merging changes here, but it is something that might be an actual issue in future so I'm leaving here at least information about this experience.

You will find the changes from this PR merged now into master under tag `4.0.0` that is also pushed to ansible galaxy.

Thank you for your time and contributions to this role! Glad to see that changes works well.

**Pull request merged**